src/modules/form/util/expressions/formula.ts

```diff
@@ -0,0 +1,70 @@
+import ExpressionEvaluationError from './ExpressionEvaluationError';
+import formulaFunctions from './formulaFunctions';
+import { parseExpression } from './parser';
+import { parseNumber } from '@/utils/numbers';
+
+const ensureNumeric = (value: any): number => {
+  if (parseNumber(value) === undefined)
+    throw new ExpressionEvaluationError('Non numeric result');
+  return value;
+};
+
+type EvalContext = Map<string, any>;
+const evaluate = (ast: any, context: EvalContext): number | undefined => {
```

Comment:
I get `N/A` no matter what if the linkIds have dashes in them. We use lots of dashes in IDs so I think that is an issue. This is what I tested in storybook:
```
        {
          "type": "TIME_OF_DAY",
          "linkId": "time-1",
          "text": "Start time"
        },
        {
          "type": "TIME_OF_DAY",
          "linkId": "time-2",
          "text": "End time"
        },
        {
          "text": "Duration: ${FORMAT_DURATION(value)}",
          "type": "DISPLAY",
          "linkId": "duration_minutes",
          "autofillValues": [
            {
              "formula": "ABS(time-1 - time-2)",
              "autofillWhen": [],
              "autofillBehavior": "ANY",
              "autofillReadonly": false
            }
          ]
        },
```

Reply:
Oh good catch, thanks! This requires some thought. Allowing operators as part of an identifier is not really compatible this (or most) parsers. 

Possible solutions:
* We could just "not do that" - if we want to use link ids in an expression, they can't have hyphens (or other operator chars).
* we could wrap each variable reference in a function with an escaped string (for json). that looks like: `"formula": "ABS($(\"end-date\") - $(\"start-date\"))"`. This starts to get hard to read and adds complexity/processing overhead but it should work
* We could technically reconfigure the parser with different operator and identifier and operator chars
* In the form refactor branch Zach built a utility that translates invalid link ids so they work with RHF. We could try and extract and adapt that solution.

Long term I'd love for us to move away from hyphens, dots, etc. Supporting those invalid chars seems limiting without providing much benefit. Ideally our link ids would also be valid JS identifiers. If we want to rip off the bandaid, I think we could write a data migration to update all to form definitions.

Reply:
honestly I think we should just go ahead and migrate them. I added a ticket here: https://www.pivotaltracker.com/story/show/187087981